mqtt-impl/src/main/java/io/streamnative/pulsar/handlers/mqtt/MQTTSubscriptionManager.java

```diff
@@ -31,8 +35,30 @@
 
     private ConcurrentMap<String, List<MqttTopicSubscription>> subscriptions = new ConcurrentHashMap<>(2048);
 
-    public void addSubscriptions(String clientId, List<MqttTopicSubscription> topicSubscriptions) {
-        this.subscriptions.computeIfAbsent(clientId, k -> new ArrayList<>()).addAll(topicSubscriptions);
+    public boolean addSubscriptions(String clientId, List<MqttTopicSubscription> topicSubscriptions) {
+        final AtomicBoolean duplicated = new AtomicBoolean(false);
+        this.subscriptions.compute(clientId, (k, v) -> {
+            if (v == null) {
```

Comment:
Looks we have race conditions here.
Because ``ConcurrentHashMap#compute`` is not an atomic operation.

```java
    default V compute(K key,
            BiFunction<? super K, ? super V, ? extends V> remappingFunction) {
        Objects.requireNonNull(remappingFunction);
        V oldValue = get(key);

        V newValue = remappingFunction.apply(key, oldValue);
        if (newValue == null) {
            // delete mapping
            if (oldValue != null || containsKey(key)) {
                // something to remove
                remove(key);
                return null;
            } else {
                // nothing to do. Leave things as they were.
                return null;
            }
        } else {
            // add or replace old mapping
            put(key, newValue);
            return newValue;
        }
    }
```

Reply:
Fixed.